internal/backend/remote-state/gcs/backend_test.go

```diff
@@ -211,6 +212,56 @@ func TestBackendWithCustomerManagedKMSEncryption(t *testing.T) {
 	backend.TestBackendStateLocks(t, be0, be1)
 }
 
+func TestBackendEncryptionKeyEmptyConflict(t *testing.T) {
+	// This test is for the edge case where encryption_key and
+	// kms_encryption_key are both set in the configuration but set to empty
+	// strings. The "SDK-like" helpers treat unset as empty string, so
+	// we need an extra rule to catch them both being set to empty string
+	// directly inside the configuration, and this test covers that
+	// special case.
+	//
+	// The following assumes that the validation check we're testing will, if
+	// failing, always block attempts to reach any real GCP services, and so
+	// this test should be fine to run without an acceptance testing opt-in.
+
+	// This test is for situations where these environment variables are not set.
+	t.Setenv("GOOGLE_ENCRYPTION_KEY", "")
+	t.Setenv("GOOGLE_KMS_ENCRYPTION_KEY", "")
+
+	backend := New()
+	schema := backend.ConfigSchema()
+	rawVal := cty.ObjectVal(map[string]cty.Value{
+		"bucket": cty.StringVal("fake-placeholder"),
+
+		// These are both empty strings but should still be considered as
+		// set when we enforce teh rule that they can't both be set at once.
+		"encryption_key":     cty.StringVal(""),
+		"kms_encryption_key": cty.StringVal(""),
+	})
+	// The following mimicks how the terraform_remote_state data source
+	// treats its "config" argument, which is a realistic situation where
+	// we take an arbitrary object and try to force it to conform to the
+	// backend's schema.
+	configVal, err := schema.CoerceValue(rawVal)
+	if err != nil {
+		t.Fatalf("unexpected coersion error: %s", err)
+	}
+	configVal, diags := backend.PrepareConfig(configVal)
+	if diags.HasErrors() {
+		t.Fatalf("unexpected PrepareConfig error: %s", diags.Err().Error())
+	}
+
+	configDiags := backend.Configure(configVal)
+	if !configDiags.HasErrors() {
+		t.Fatalf("unexpected success; want error")
+	}
+	gotErr := configDiags.Err().Error()
+	wantErr := `can't set both encryption_key and kms_encryption_key`
+	if !strings.Contains(gotErr, wantErr) {
+		t.Errorf("wrong error\ngot: %s\nwant substring: %s", gotErr, wantErr)
+	}
```

Comment:
Thanks for sharing this example of how to test for errors! I'll take some inspiration from it. I always thought that testing errors wasn't possible because the existing tests used `TestBackendConfig` and [errors cause fails there](https://github.com/hashicorp/terraform/blob/c77898c90d95cd1a44970af6af302024f73e56cb/internal/backend/testing.go#L47-L50).

Reply:
Testing for errors using substrings _can_ be tricky in some cases, because sometimes what we're testing against was already formatted for presentation to humans and so contains annoyances like word-wrapping. :confounded: 

But in this particular case we are working directly with the `tfdiags.Diagnostics` object, rather than the formatted result generated from it, and so it's more reasonable (but still a bit sneaky) to do what I did here, of relying on `tfdiags`'s ability to treat a `tfdiags.Diagnostics` as a normal Go `error`, whose `Error()` implementation is considerably more simplistic than our UI-level diagnostics formatter, and in particular never introduces line wrapping.

I mention all of that only because I've been tripped up in the past by trying to use patterns _like_ this at higher levels of abstraction, but thankfully when we're in the backend packages we're not exposed to such hazards.  :grinning:  This approach of testing substrings of `diags.Err().Error()` is quite prevalent in this codebase and so if we ever were to break it we'd break a lot of tests all at once.